### PR TITLE
Text Selection API for Input and TextArea

### PR DESCRIPTION
**Pull request by sjelin**
See #799.  Followed documentation at http://www.whatwg.org/specs/web-apps/current-work/#textFieldSelection

Pretty straight forward addition.  A few concerns:

1. I ended up duplicating the code in order to put it into both Input and Text Area
2. When the content of the text box changes, presumably the selection is also supposed to change.  Consider, for example, if the content of the box is empties, and the selection is supposed to be of characters 4 through 10.  Clearly that selection is now impossible.  However, I couldn't really find any documentation on this point (though it's possible I don't know where to look).  Chrome moves the cursor to the end of the text box, which seems reasonable, but since I couldn't find it actually documented anywhere I just did some basic sanitizing to make sure the selection is always within the bounds of the actual text.
3. ~~Sometimes `input` tags don't support the selection API (e.g. `type=radio`).  In these cases I throw `throw new core.DOMException(core.INVALID_STATE_ERR)`, because chrome throws an `InvalidStateError` and this was the most similar thing I could find.  No idea if that's the correct thing to throw though.  Would appreciate someone who knows more taking a look.~~ Never mind, the W3C tests seem to imply that this is the correct thing to throw.
4.  ~~On my machine, the automated tests stall at `running jsdom/index.js jsdom/index`.  They did this before I made my changes, so I'm pretty sure it's not my fault, but at the same time I feel kind of queasy sending a pull request without having the tests finish.~~ This stopped happening when I rebased against 1.0 (yay?)
5.  ~~On that note, I haven't written any tests.  I did some manual testing, so I know it everything works.  I don't know if that's enough for you guys.~~ See the discussion below

Also, the last three commits are all just small bug fixes.  The "merge" was just that I stupidly tried to amend my first bug fixing commit after pushing it to github.

### CODE REVIEW

File: lib/jsdom/level2/html.js [hunk 1]
```diff
@@ -1025,6 +1027,9 @@ define('HTMLInputElement', {
       else {
         this.setAttribute('value', val);
       }
+      if(this._allowSelection()) {
+        this.selectionStart = this.selectionStart; // Ensure selection doesn't go past end of new value
```
Comment (fluffybunnies) [inline lib/jsdom/level2/html.js]: this._selectionStart ?

**Comment by windyarts** [comment]
Will you merge this?

**Comment by domenic** [comment]
It's still missing tests :(

**Comment by sjelin** [comment]
Wow, totally forgot about it.  I'll deal with it this weekend, sorry

**Comment by benmosher** [comment]
Looking forward to this! 👍

**Comment by kentmw** [comment]
Also looking forward to this. :)

**Comment by yaycmyk** [comment]
@sjelin Looks like this needs a rebase, when you get the chance.

**Comment by domenic** [comment]
I can do the rebase if someone writes tests

**Comment by yaycmyk** [comment]
@domenic I'll take a crack at this - want me to submit as a new PR since I can't add onto this one?

**Comment by domenic** [comment]
That sounds great! Just be sure to read CONTRIBUTING.md about how to turn on web platform tests that exist already, and how new tests should be written in that style. Feel free to bug us in #jsdom on freenode IRC if you have any questions.

**Comment by sjelin** [comment]
It's a thanksgiving miracle!  Also, sorry guys, I got very sick shortly after sending this PR and then switched jobs so it was very low priority. @yanovich let me know if you have any questions.  I also give you permission to take full authorship if that makes your life easier.

**Comment by domenic** [comment]
Superceded by #1305 :)

**Pull request closed without merging**